Cholesky backpropagation

alexggmatthews opened the pull request:
This implements a gradient op for the Cholesky op as requested and discussed in #367 .

The code was written by myself and Dr. James Henman (user name @jameshensman ).

The implementation is a blocked cholesky backpropagation as advocated by Dr. Iain Murray in this recent paper http://arxiv.org/abs/1602.07527

New unit tests are included. 

A lot of work has gone into the implementation of the algorithm and I therefore believe the code has value to the community. Less work so far has gone in to tightly integrating it with TensorFlow code base and I am looking for help and advice from the TensorFlow team. As such it should be regarded as a work in progress.  @martinwicke suggested the best path forward was to submit this pull request.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

girving commented:
@rmlarsen: Want to take a look?

jameshensman commented:
I signed the CLA.

alexggmatthews commented:
I signed the CLA. All authors have now signed.

rmlarsen commented:
Thank Alex, I'm reviewing the code now.

On Mon, Mar 14, 2016 at 1:09 AM, Alexander G. de G. Matthews <
notifications@github.com> wrote:

> I signed the CLA. All authors have now signed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1465#issuecomment-196192589>
> .
>

rmlarsen commented on tensorflow/python/user_ops/user_ops.py:
```diff
@@ -26,3 +26,11 @@
+
+@ops.RegisterGradient("Cholesky")
+def _cholesky_grad(op, grad):
+  return ( cholesky_grad( op.outputs[0] , grad ) )
```
Remove outer parens and space inside inner parens.

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+//Copyright (c) 2016, Alexander G. de G. Matthews and James Hensman. All rights reserved.
```
I think this code should go in tensorflow/core/kernels/cholesky_grad.cc.

rmlarsen replied:
# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
You need to remove this copyright notice and replace it with the standard Apache license used in Tensorflow:

# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

rmlarsen replied:
You need to remove this copyright notice and replace it with the standard Apache license header.

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/framework/types.h"
```
Please put the op registration and docstring in tensorflow/core/ops/linalg_ops.cc.
Please put one argument per line (newline before each '.').

I would recommend not using the notation from Murray's paper to name the input and output arguments. Perhaps name them something like "cholesky", "cholesky_grad", and "matrix_grad". 

Please make your docstring a bit more detailed (see the registration of cholesky in ops/linalg_ops.cc for example). Also, get rid of references to undefined entities "f" and "l"; I think it suffices to say that the function calculates the gradient w.r.t. to the input matrix, given the gradient w.r.t. its Cholesky factorization.

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+    const ConstMatrixMap input_matrix_l_bar(input_tensor_l_bar.flat<T>().data(), input_tensor_l_bar.dim_size(0), input_tensor_l_bar.dim_size(1));
+    MatrixMap output_matrix(output_tensor->template flat<T>().data(), input_tensor_l.dim_size(0), input_tensor_l.dim_size(1) );    
+
+    const lcl_size_t N = input_matrix_l.rows();
```
Variable names should be lowercase. Same everywhere.

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+    }
+    
+        
+    //TODO: what to do if input_matrix_l isn't lower triangular?
```
You can get rid of this TODO. The op should ignore any contents above the diagonal. The docstring should explicitly mention this.

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+    using MatrixMap = Eigen::Map<Matrix>;
+    using ConstRef = Eigen::Ref<const Matrix>;
+    using Ref = Eigen::Ref<Matrix>;
+    using lcl_size_t = int;
```
Get rid of this typedef and use int64. Same everywhere.

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+
+    void Compute(OpKernelContext* context) override {
+
+    const Tensor& input_tensor_l = context->input(0); ;
```
maybe name these input_matrix_l and input_matrix_grad, so it is clearer without having to know the notation in the paper.

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+    output_matrix = input_matrix_l_bar.template triangularView<Eigen::Lower>();
+
+    
+    for ( lcl_size_t Ji = (N-NB+1) ; Ji>(1-NB); Ji-= NB )   
```
A few general code style guidelines:
  * no space inside parens. 
  * space after comma and on both sides of binary operators. 
  * curly braces at the end of line,
  * use descriptive variable names.

You can see the full set of guidelines here: http://google.github.io/styleguide/cppguide.html

I would change the variable names to:

NB -> kMaxBlockSize
Ji ->  block_end (and redefined to point at the first element to the right of the current panel as is common with iterators in C++)
J -> block_begin
JB -> block_size

I guess this was "translated" from one-based to zero-based indexing and the subtraction of 1 everywhere makes the code harder to read. To improve readability I would suggest using something like the following:

for (int64 block_end = n; block_end > 0; block_end -= kMaxBlockSize) {
  const int64 block_begin = std::max(0ll, block_end - kMaxBlockSize);
  const int64 block_size = block_end - block_begin;
  const int64 trailing_size = n - block_end;
  output_matrix.block(block_end, block_begin, trailing_size, block_size) = ...
  output_matrix.block(block_begin, block_begin, block_size, block_size) = ...
  etc.

rmlarsen commented on tensorflow/python/kernel_tests/cholesky_grad_op_test.py:
```diff
@@ -0,0 +1,154 @@
+
+from scipy.linalg.blas import dsymv
+
+#Reference Python implementation ported from Iain Murray Matlab code. Blocked version.
```
While these tests were probably very helpful for developing the code, they are not needed as unit tests when checking this into TensorFlow. The standard approach is to have a test that compares the analytical gradients against a finite difference approximation. This functionality is provided by the tf.test.compute_gradient_error() method. You can find plenty of examples in tensorflow/python/ops/*_grad_test.py.

rmlarsen commented on tensorflow/python/user_ops/user_ops.py:
```diff
@@ -26,3 +26,11 @@
   """Example of overriding the generated code for an Op."""
   return gen_user_ops._fact()
+
+@ops.RegisterGradient("Cholesky")
```
This should go in tensorflow/python/ops/linalg_grad.py. You might look at how ReluGrad is registered in nn_grad.py as an example.

rmlarsen commented on tensorflow/python/user_ops/user_ops.py:
```diff
@@ -26,3 +26,11 @@
+def _cholesky_grad(op, grad):
+  return ( cholesky_grad( op.outputs[0] , grad ) )
+
+@ops.RegisterShape("CholeskyGrad")
```
This should go in linalg_ops.py

rmlarsen commented on tensorflow/core/user_ops/chol_grad.cc:
```diff
@@ -0,0 +1,115 @@
+        lcl_size_t J = std::max<lcl_size_t>(1, Ji);
+        lcl_size_t JB = NB - (J - Ji);
+    
+        output_matrix.block( J+JB-1, J-1, N - (J+JB-1), JB) = input_matrix_l.block( J-1, J-1, JB, JB ).adjoint().template triangularView<Eigen::Upper>().solve( output_matrix.block( J+JB-1, J-1, N - (J+JB-1), JB ).adjoint() ).adjoint();
```
Please obey the max 80 character line length rule. Here you can make the algorithm clearer by putting each of the main steps on a separate line (breaking before "."). Same everywhere.

rmlarsen commented:
Hi Alex & James,

Thanks for your contribution, this will be really nice to have in
TensorFlow. I left a number of specific comments on the pull request.

Best,
   Rasmus

On Mon, Mar 14, 2016 at 8:49 AM, Rasmus Larsen <rmlarsen@google.com> wrote:

> Thank Alex, I'm reviewing the code now.
>
> On Mon, Mar 14, 2016 at 1:09 AM, Alexander G. de G. Matthews <
> notifications@github.com> wrote:
>
>> I signed the CLA. All authors have now signed.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/tensorflow/tensorflow/pull/1465#issuecomment-196192589>
>> .
>>
>
>

alexggmatthews commented:
Hi Rasmus,
Thanks for your comments that is a big help. We'll get on it!
Best,
Alex

alexggmatthews commented:
Hi @rmlarsen,
I am about half way through your comments. I will keep you posted and let you know when I am done.
Alex

rmlarsen commented:
Sounds great. I'm back from a short vacation, so please let me know when the code is ready for another look.

Rasmus

alexggmatthews commented:
Hi @rmlarsen,
I believe I have now addressed all of your comments so I think it would be good for you to have another look please. 

In terms of complying with the coding standard I ran the lint python script they suggest that can be found here https://raw.githubusercontent.com/google/styleguide/gh-pages/cpplint/cpplint.py

Thank you for your lucid comments that I feel have definitely improved both the style and content of the code. 

It is a public holiday here in the UK tomorrow (Friday) and Monday but I will be back at my desk on Tuesday. 

Alex

rmlarsen commented on tensorflow/core/kernels/cholesky_grad.cc:
```diff
@@ -0,0 +1,164 @@
+                   .adjoint()
+                   .template triangularView<Eigen::Upper>()
+                   .solve(output_matrix
+                   .block(block_end, block_begin, trailing_size, block_size)
```
Indent here to reflect the block structure.

alexggmatthews replied:
Could you be a bit more specific about what you want here please? Thanks, Alex

rmlarsen replied:
I think running clang-format is the easiest way to address this.

rmlarsen commented on tensorflow/core/kernels/cholesky_grad.cc:
```diff
@@ -0,0 +1,164 @@
+                   -= (output_matrix
+                   .block(block_end, block_begin, trailing_size, block_size)
+                   .adjoint()
+                   *input_matrix_l.block(block_end, block_begin, trailing_size,
```
space around binary operators. Same everywhere.

rmlarsen commented on tensorflow/python/kernel_tests/cholesky_op_test.py:
```diff
@@ -87,6 +88,26 @@ def testEmpty(self):
+      for shape in shapes:             
+        for dtype in dtypes:
+          x = tf.constant( np.random.rand(shape[0]) )
+          R = tf.constant( np.random.randn(shape[0],shape[1])  )
```
Space after comma. Get rid of spaces inside parens. Same everywhere.

alexggmatthews replied:
Where is python style documented please?

rmlarsen replied:
The tensorflow python code follows that PEP8 style guide. https://www.python.org/dev/peps/pep-0008/

alexggmatthews replied:
@rmlarsen @vrv is there any automatic formatting software that is used on TensorFlow python code, please? i.e is there a Python equivalent of running clang-format ?

vrv replied:
Unfortunately we don't run it all that much on the python code (the clang-format one is more or less enforced internally), but you might want to try pyformat --yapf (which is what I use, with two space indent rather than 4).

https://github.com/google/yapf

rmlarsen commented on tensorflow/python/kernel_tests/cholesky_op_test.py:
```diff
@@ -22,6 +22,7 @@
 from six.moves import xrange  # pylint: disable=redefined-builtin
 import tensorflow as tf
 
+from IPython import embed
```
Delete this?

alexggmatthews replied:
My apologies. That should not have made it in.

rmlarsen commented on tensorflow/core/kernels/cholesky_grad.cc:
```diff
@@ -0,0 +1,164 @@
+      const int64 block_begin = std::max(0ll, block_end - kMaxBlockSize);
+      const int64 block_size = block_end - block_begin;
+      const int64 trailing_size = kMatrixSize - block_end;
+      output_matrix.block(block_end, block_begin, trailing_size , block_size)
```
Looking at this again,  I'm thinking it might be helpful to add a little ASCII art comment showing the partitioning corresponding to the "level3partition" function in the paper and annotate the steps with comments indicating the mathematical operation, for example // \bar{C} = \bar{C} * inv(D) for this line.

Alternatively/additionally, define local variables for the various blocks with consistent names, such  as

auto C_bar =  output_matrix.block(block_end, block_begin, trailing_size , block_size);
auto D = input_matrix_l.block(block_begin, block_begin, block_size, block_size);
C_bar = D.adjoint().template triangularView<Eigen::Upper>().solve(C_bar).adjoint()).adjoint();

etc.

alexggmatthews replied:
OK I will have a look at defining local variables.

alexggmatthews replied:
This was a particularly helpful comment that significantly improved the clarity of the code.  I was initially concerned about using temps for speed reasons but there weren't any when we did it in the way suggested here.

rmlarsen commented on tensorflow/core/kernels/cholesky_grad.cc:
```diff
@@ -0,0 +1,164 @@
+  void CholeskyGradUnblocked(const ConstRef l_block, Ref grad_block) {
+    const int64 kMatrixSize = l_block.rows();
+    for (int64 k = kMatrixSize-1; k >= 0; k--) {
+      grad_block(k, k) -= (l_block
```
This might also benefit from defining the blocks as temporaries.

alexggmatthews replied:
OK I will have a look at defining local variables.

rmlarsen commented on tensorflow/core/kernels/cholesky_grad.cc:
```diff
@@ -0,0 +1,164 @@
+      output_matrix.block(block_end, block_begin, trailing_size , block_size)
+                   = input_matrix_l
+                   .block(block_begin, block_begin, block_size, block_size)
+                   .adjoint()
```
In general: You don't need to break the line before _every_ '.' as long as the expressions fit on a 80 char line and it's possible to make the indentation reflect the block structure.

vrv replied:
(might be worth running clang-format over the changed lines -- that's what we do internally)

rmlarsen replied:
Hi Alex,

Thanks for the much improved version. I left a few additional comments on
the PR. As Vijay suggests, it might be a good idea to run clang-format on
your .cc and .h files.

Rasmus

On Mon, Mar 28, 2016 at 4:27 PM, Vijay Vasudevan <notifications@github.com>
wrote:

> In tensorflow/core/kernels/cholesky_grad.cc
> <https://github.com/tensorflow/tensorflow/pull/1465#discussion_r57650974>:
>
> > +    // Algorithm only depends on lower triangular half on input_matrix_grad.
> > +    output_matrix = input_matrix_grad.template triangularView<Eigen::Lower>();
> > +
> > +    const int64 kMatrixSize = input_matrix_l.rows();
> > +    const int64 kMaxBlockSize = 32;
> > +
> > +    for (int64 block_end = kMatrixSize;
> > +               block_end > 0;
> > +               block_end -= kMaxBlockSize) {
> > +      const int64 block_begin = std::max(0ll, block_end - kMaxBlockSize);
> > +      const int64 block_size = block_end - block_begin;
> > +      const int64 trailing_size = kMatrixSize - block_end;
> > +      output_matrix.block(block_end, block_begin, trailing_size , block_size)
> > +                   = input_matrix_l
> > +                   .block(block_begin, block_begin, block_size, block_size)
> > +                   .adjoint()
>
> (might be worth running clang-format over the changed lines -- that's what
> we do internally)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1465/files/37f2bb7b24fd41dec1b5d5d1c2dc1b4c99ef4de7#r57650974>
>

rmlarsen commented on tensorflow/python/kernel_tests/cholesky_op_test.py:
```diff
@@ -87,6 +88,26 @@ def testEmpty(self):
+          x = tf.constant( np.random.rand(shape[0]) )
+          R = tf.constant( np.random.randn(shape[0],shape[1])  )
+          e = tf.matmul( tf.diag( x ), R )
+          K = tf.matmul( e, tf.transpose(e) ) / shape[0] #This gives a positive definite matrix.
```
Why not just construct A^T * A for some random A?

alexggmatthews replied:
This was to make the test run faster - the full matrix multiply takes 44 seconds on my machine built under opt as opposed to 22 seconds with what I currently have. What do you think?

rmlarsen replied:
OK, generally we want individual tests to run in 30-45 seconds without optimization. You might want to break your test into multiple ones instead of looping over the combination of all parameters within a single test.

alexggmatthews commented:
Hi @rmlarsen,
My apologies for the slow reply. Thank you again for your comments. I've left some replies and some questions at the relevant lines. I will be looking working on defining local variables in the two main loops while I wait for your reply. Thanks again,
Alex

ebrevdo commented:
Would be great if you could make this batch cholesky grad, for which cholesky grad is a simple special case requiring a reshape.

alexggmatthews commented:
Hi @ebrevdo , I think it is best if that is done in a separate PR. This one is already getting rather large and I'm mindful of specification creep. Good suggestion though.

rmlarsen commented on tensorflow/core/kernels/cholesky_grad.cc:
```diff
@@ -0,0 +1,170 @@
+      | R D  |
+      \ B C  /
+      
+      `bar' denotes a block of the derivative matrix.
```
Perhaps be a bit more verbose: "Variables names representing the derivative matrix have a trailing '_bar'."

rmlarsen commented on tensorflow/python/kernel_tests/cholesky_op_test.py:
```diff
@@ -58,35 +57,85 @@ def testBatch(self):
+  def testOneBlockMatrices(self):
+    np.random.seed(0)  
+    shapes = self.getShapes([self._backprop_block_size+1])
+    self.runFiniteDifferences(shapes,dtypes=(tf.float32, tf.float64),scalarTest=True)      
```
Nit: Space after comm. Same everywhere.

rmlarsen commented on tensorflow/python/kernel_tests/cholesky_op_test.py:
```diff
@@ -58,35 +57,85 @@ def testBatch(self):
+  _backprop_block_size = 32   
+  
+  def getShapes(self,shapeList):
+    return ( (elem, int(np.floor(1.2*elem))) for elem in shapeList )
```
Nit: No spaces inside parens. Same everywhere.

rmlarsen commented:
Hi Alex, I think the code looks good now. I had a few nits about the formatting of the python code. I agree that implementing batching can be done in a followup PR.

alexggmatthews commented:
Hi @rmlarsen,
I have added the longer comment about _bar you asked for. 

@jameshensman has spent some time using flake8 on the unit test to ensure pep8 compliance which also covers your points on this. It turns out some of the rest of the script wasn't pep8 compliant so he fixed that too. 

I you agree, I think that therefore means we are good to go. What happens next? 

Thanks for all your help, 
@alexggmatthews

rmlarsen commented:
On Tue, Apr 5, 2016 at 1:19 AM, Alexander G. de G. Matthews <
notifications@github.com> wrote:

> Hi @rmlarsen <https://github.com/rmlarsen>,
> I have added the longer comment about _bar you asked for.
>
> @jameshensman <https://github.com/jameshensman> has spent some time using
> flake8 on the unit test to ensure pep8 compliance which also covers your
> points on this. It turns out some of the rest of the script wasn't pep8
> compliant so he fixed that too.
>
> I you agree, I think that therefore means we are good to go. What happens
> next?
>

Thanks for all your work - it looks nice. I will test your PR internally
and then we can to merge it. But first, it would be preferable if you could
squash your work into a single commit. You can either use the script
at http://rebaseandsqua.sh/
or follow these instructions: https://davidwalsh.name/squash-commits-git

Best,
   Rasmus

Thanks for all your help,
> @alexggmatthews <https://github.com/alexggmatthews>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1465#issuecomment-205716282>
>

rmlarsen commented:
Jenkins, test this please.

alexggmatthews commented:
Hi Rasmus, I agree it looks nice. Very satisfying. Thanks for all your work too. I've done the squash so we now have just the one commit as requested. Do we need to keep merging master to this branch while we wait for Jenkins so that it doesn't get out of date or is that dealt with automatically? 
Alex

rmlarsen commented:
Hi Alex,

I suspect you have to merge manually for now. Perhaps @martinwicke knows?

vrv commented:
Now that the squash/merge is done, we can test it once more and then merge.
@tensorflow-jenkins: test this please.

(alternatively, we could have squashed/merged ourselves).

vrv commented:
Looks like some of the tests are timing out -- can you figure out which test is taking so long and try to reduce the entire test suite to run in < 30 seconds?

vrv commented:
Also, the mac test failed with:

INFO: From Testing //tensorflow/python:cholesky_op_test:
==================== Test output for //tensorflow/python:cholesky_op_test:
.F..........
======================================================================
FAIL: testSmallMatrices (__main__.CholeskyGradTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/tensorflow-pull-requests-mac/bazel-out/local_darwin-opt/bin/tensorflow/python/cholesky_op_test.runfiles/tensorflow/python/kernel_tests/cholesky_op_test.py", line 100, in testSmallMatrices
    self.runFiniteDifferences(shapes)
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/tensorflow-pull-requests-mac/bazel-out/local_darwin-opt/bin/tensorflow/python/cholesky_op_test.runfiles/tensorflow/python/kernel_tests/cholesky_op_test.py", line 138, in runFiniteDifferences
    self.assertLess(error, 1e-3)
AssertionError: 0.0011898875 not less than 0.001

alexggmatthews commented on tensorflow/python/kernel_tests/cholesky_op_test.py:
```diff
@@ -88,5 +88,55 @@ def testEmpty(self):
+            R = tf.constant(np.random.randn(shape[0], shape[1]), dtype)
+            e = tf.mul(R, x)
+            K = tf.matmul(e, tf.transpose(e)) / shape[0]  # K is posdef
+            y = tf.reduce_mean(tf.cholesky(K))
```
@rmlarsen @vrv The speed problem was coming from the fact that for high dimensional outputs y testing the whole cholesky matrix was prohibitively slow. I fixed this by testing a scalar function of the cholesky matrix for the large test cases. This is a realistic use case in practice (e.g computing the log determinant of a psd matrix).  The whole cholesky test on my desktop now takes 0.7 seconds run directly and a bit slower run from bazel. We still have coverage for the whole matrix case on the smaller test matrices. I've re-squashed this hence why it doesn't appear in the diff.

alexggmatthews commented:
I am looking into the mac problem but I don't have one in my laboratory so I am going to liaise with @jameshensman . I guess there must be slight numerical differences between the two.

rmlarsen commented:
Hi Alex, Given that the OS-X test fails due to a tiny numerical discrepancy, I'd say that increasing the test tolerance by a factor of 2, say, is an acceptable fix.

alexggmatthews committed:
Pull request #1465. Implements reverse mode differentiation of the cholesky decomposition as an op in TensorFlow.

alexggmatthews commented:
OK @vrv @rmlarsen we should be good to go again. Please can you re-run the tests.

vrv commented:
@tensorflow-jenkins: test this please

alexggmatthews commented:
@vrv @rmlarsen @jameshensman it looks like it is all working.

rmlarsen commented:
Woohoo!

alexggmatthews commented:
Excellent!

rmlarsen commented:
Hi Alex,

Would you have time & interest in extending the Cholesky gradient code to
the batch case?

Thanks,
   Rasmus

On Fri, Apr 8, 2016 at 1:01 AM, Alexander G. de G. Matthews <
notifications@github.com> wrote:

> Excellent!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1465#issuecomment-207289948>
>

alexggmatthews commented:
Hi Rasmus, it does sound interesting but I'm now very near finishing my PhD thesis so I think I will have to mainly concentrate on that for the time being. I do have a small PR #1891 that I would be happy to help through. Best, Alex
